File: bindings/nostr-sdk-ffi/python/examples/whitelist.py
import asyncio
from nostr_sdk import PublicKey, ClientBuilder, Filter, Kind, init_logger, LogLevel, AdmitPolicy, AdmitStatus, Event, \
    uniffi_set_event_loop
from datetime import timedelta

class WoT(AdmitPolicy):
    def __init__(self):
        self.allowed_public_keys = set()

    def allow(self, pk: PublicKey):
        self.allowed_public_keys.add(pk)

    async def admit_event(self, relay_url: str, subscription_id: str, event: Event) -> AdmitStatus:
        if event.author() in self.allowed_public_keys:
            return AdmitStatus.SUCCESS
        else:
            return AdmitStatus.REJECTED

async def main():
    uniffi_set_event_loop(asyncio.get_running_loop())

    # Init logger
    init_logger(LogLevel.INFO)

    whitelisted_public_key = PublicKey.parse("npub1l2vyh47mk2p0qlsku7hg0vn29faehy9hy34ygaclpn66ukqp3afqutajft")
    not_whitelisted_public_key = PublicKey.parse("npub1xtscya34g58tk0z605fvr788k263gsu6cy9x0mhnm87echrgufzsevkk5s")

    wot = WoT()
    wot.allow(whitelisted_public_key)

    # Init client
    client = ClientBuilder().admit_policy(wot).build()
    await client.add_relay("wss://relay.damus.io")
    await client.connect()

    # Get events
    f = Filter().authors([whitelisted_public_key, not_whitelisted_public_key]).kind(Kind(0))
    events = await client.fetch_events(f, timedelta(seconds=10))
    print(f"Received {events.len()} events")


if __name__ == '__main__':
    asyncio.run(main())

File: sdk/examples/whitelist.rs
// Copyright (c) 2022-2023 Yuki Kishimoto
// Copyright (c) 2023-2025 Rust Nostr Developers
// Distributed under the MIT software license

use std::collections::HashSet;
use std::time::Duration;

use nostr_sdk::prelude::*;

#[derive(Debug, Default)]
struct WoT {
    allowed_public_keys: HashSet<PublicKey>,
}

impl AdmitPolicy for WoT {
    fn admit_event<'a>(
        &'a self,
        _relay_url: &'a RelayUrl,
        _subscription_id: &'a SubscriptionId,
        event: &'a Event,
    ) -> BoxedFuture<'a, Result<AdmitStatus, PolicyError>> {
        Box::pin(async move {
            if self.allowed_public_keys.contains(&event.pubkey) {
                return Ok(AdmitStatus::success());
            }

            Ok(AdmitStatus::rejected("Not in whitelist"))
        })
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();

    let mut wot = WoT::default();

    // Allow public key
    let allowed_public_key =
        PublicKey::from_bech32("npub1drvpzev3syqt0kjrls50050uzf25gehpz9vgdw08hvex7e0vgfeq0eseet")?;
    wot.allowed_public_keys.insert(allowed_public_key);

    // Init client
    let client = Client::builder().admit_policy(wot).build();
    client.add_relay("wss://relay.damus.io").await?;
    client.connect().await;

    let not_in_whitelist_public_key =
        PublicKey::from_bech32("npub1xtscya34g58tk0z605fvr788k263gsu6cy9x0mhnm87echrgufzsevkk5s")?;

    // Get events from all connected relays
    let filter = Filter::new()
        .authors([allowed_public_key, not_in_whitelist_public_key])
        .kind(Kind::Metadata);
    let events = client.fetch_events(filter, Duration::from_secs(10)).await?;
    println!("Received {} events.", events.len());

    assert_eq!(events.len(), 1);
    assert_eq!(events.first().unwrap().pubkey, allowed_public_key);

    for event in events.into_iter() {
        println!("{}", event.as_json());
    }

    Ok(())
}
